docs/writing-controllers.md

```diff
@@ -0,0 +1,1489 @@
+
+```typescript
+const name = 'FooController';
+
+export type MessageReceived = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+✅ **Type name now matches controller name + event identifier + "Event"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerMessageReceivedEvent = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+## Define and export a type union for internal action types
+
+A controller should define and export a type union that holds all of its actions. This type union makes it easy to define the type for the controller's messenger.
+
+The name of this type should be `${ControllerName}Actions`.
+
+This type should be only passed to `RestrictedControllerMessenger` as the 2nd type parameter. It should _not_ be included in its 4th type parameter, as that is is used for external events.
+
+🚫 **`FooController['type']` is passed as the 4th type parameter**
+
+```typescript
+export type FooControllerActions =
+  | FooControllerUpdateCurrencyAction
+  | FooControllerUpdateRatesAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  FooControllerActions,
+  never,
+  FooControllerActions['type'],
+  never
+>;
+```
+
+✅ **`never` is passed as the 5th type parameter (assuming no external events)**
+
+```typescript
+export type FooControllerActions =
+  | FooControllerUpdateCurrencyAction
+  | FooControllerUpdateRatesAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  FooControllerActions,
+  never,
+  never,
+  never
+>;
+```
+
+## Define and export internal event types as one
+
+A controller should define and export a type union that holds all of its events. This type union makes it easy to define the type for the controller's messenger.
+
+The name of this type should be `${ControllerName}Events`.
+
+This type should be only passed to `RestrictedControllerMessenger` as the 3rd type parameter. It should _not_ be included in its 5th type parameter, as that is is used for external events.
+
+🚫 **`FooControllerEvents['type']` is passed as the 5th type parameter**
+
+```typescript
+export type FooControllerEvents =
+  | FooControllerMessageReceivedEvent
+  | FooControllerNotificationAddedEvent;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  never,
+  FooControllerEvents,
+  never,
+  FooControllerEvents['type']
+>;
+```
+
+✅ **`never` is passed as the 5th type parameter (assuming no external events)**
+
+```typescript
+export type FooControllerEvents =
+  | FooControllerMessageReceivedEvent
+  | FooControllerNotificationAddedEvent;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  never,
+  FooControllerEvents,
+  never,
+  never
+>;
+```
+
+## Define external action types as one, but do not export them
```

Comment:
For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670856411

```suggestion
## Define, but do not export, a type union for external action types
```